libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+
+using namespace std;
+using namespace solidity::frontend;
+
+namespace {
+
+bool nodeSet(FunctionCallGraphBuilder::Node const& _node)
+{
+	if (auto node = std::get_if<FunctionCallGraphBuilder::SpecialNode>(&_node))
+		return *node != FunctionCallGraphBuilder::SpecialNode::Unset;
+
+	return true;
+}
+
+}
+
+namespace solidity::frontend
+{
+
+
+shared_ptr<FunctionCallGraphBuilder::ContractCallGraph> const FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	m_contract = &_contract;
+
+	m_graph = make_shared<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::CreationRoot;
+	m_currentDispatch = SpecialNode::CreationDispatch;
+	visitConstructor(
+		_contract,
+		_contract.annotation().linearizedBaseContracts.cbegin() + 1,
+		_contract.annotation().linearizedBaseContracts.cend()
+	);
+	m_currentNode = SpecialNode::Unset;
+	m_currentDispatch = SpecialNode::RuntimeDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+	}
+
+	// Add all CreationDispatch calls to the RuntimeDispatch as well
+	for (auto node: m_graph->edges[SpecialNode::CreationDispatch])
+		add(SpecialNode::RuntimeDispatch, node);
+
+	// Add all external functions to the RuntimeDispatch
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
```

Comment:
This is really annoying. is there samething like `auto& [_, functionType]:`?

Reply:
Unfortunately, nothing like that made it into the standard, yet - there may be either ``auto& [std::ignore, functionType]`` or ``auto& [_, functionType]`` or ``auto& [*, functionType]`` some time in the future, but for the time being there's no really nice way to do this.

However, C++20 and ranges-v2 does have ``std::ranges::views::values`` - with that this could become
```
for (auto& functionType: _contract.interfaceFunctionList() | std::ranges::views::values)
```
So maybe we should go for ranges-v3 :-)?

Reply:
Done.